### ISSUE TITLE
Locking to 40 FPS with ASW disabled (lowest settings), depth scale constantly changing, and clipping occurs in vehicles

### ISSUE THREAD

**Issue by aaronth07** [LukeRoss00/gta5-real-mod#376]
I thought I would combine these ~~3~~ 4 issues i'm having into one post so I don't clutter the issues page.

Required info:

- **CPU**: i7-7700k
- **GPU**: 1070 Ti
- **VR headset model**: Rift S
- **game version and build**: 1.0.2245.0
- **release of the mod**: Release 7
- **did you follow all the Quick setup steps carefully, in particular step 4?**: Yes
- **which settings did you choose (H/M/L) when running the batch file RealConfig.bat?**: M
- **did the batch file report any errors?**: No
- **did you change the graphics settings manually?**: Yes, I have changed the texture setting, ambient occlusion setting, and frame scaling setting when playing normally, but have tried different settings when testing. All issues still happened with default settings.
- **are you using any other mods?**: No

**First issue: I'm always locked to 40 FPS no matter what my graphics settings are at.** 

I have tried disabling ASW (Asynchronous Space Warp) through OTT (Oculus Tray Tool), but that only makes it worse since, well, ASW is disabled. I don't think disabling ASW through OTT actually disables the 40 FPS lock, just ASW. And as I said, this happens with all graphics settings, even at lowest everything and frame scaling at 0.75. And I have V-Sync disabled in-game and NVIDIA Control Panel is set to use the game settings. 

*EDIT: After more testing this only happens with ASW disabled. When enabling ASW through the OTT I never lock to 40. However, I still have issues where the game will stutter and jump between 60 and 80 FPS even at the 0.5 frame scaling. What's weird is that this will happen the same on 0.5 frame scaling and 2.0 frame scaling, indicating it's not necessarily a GPU issue.*

**Second issue: The depth scale is constantly changing when moving.**

By "depth scale", I mean the scale of the objects (how big they appear). For example, walking close to a car, and then backing up, will have the car go from normal sized, to toy sized, to giant sized, dependant on movement and look direction. And the scale actually changes back and forth WHILE moving, it's very weird. This also happens with NPCs and cars in the distance, or coming close to me. And the actual model in-game is obviously not changing size at all, just the depth scale.

**Third issue: Clipping occurs when inside a vehicle.**

An image is worth a thousand words right?
![image](https://user-images.githubusercontent.com/22606764/125021037-80005900-e03f-11eb-9311-42289db69343.png)
This happens consistently in every vehicle. Both the left door and the roof clips. Resetting the view does nothing unless I position myself in such a way that the my position after resetting the view is offset enough from my normal position that my normal position no longer clips (basically moving my head up and to the left, resetting the view, and then going back to my normal position). Except that also means I am very off center from the character. 

**EDIT: Fourth issue (just noticed after writing this): The weapon model has a ghosting effect**

Basically, when looking around, the weapon (and presumably hands) model "ghosts". This happens even with ASW turned off in the OTT. I tried recording a video of my desktop but it didn't show in the game window and i'm not sure how to record native VR output. Also, after this issue, I tried uninstalling OTT and instead using the Oculus Debug Tool to disable ASW (I thought maybe OTT wasn't working), but all of these issues occur even with disabling ASW through the debug tool.

Files:

[ScriptHookV.log](https://github.com/LukeRoss00/gta5-real-mod/files/6788625/ScriptHookV.log)
[asiloader.log](https://github.com/LukeRoss00/gta5-real-mod/files/6788626/asiloader.log)
[RealVR.ini](https://pastebin.com/FW5wbQKM)

**Comment by LukeRoss00** [comment]
Well, that's a lot of information! First of all, don't start messing with the graphics settings before reading the README. For instance, you're talking about setting Frame Scaling to 0.5 or 0.75 when [here](https://github.com/LukeRoss00/gta5-real-mod#advanced-tweaking-and-hotkeys) I clearly state that it must be 2.5 or 2 for the mod to work correctly.

So, issue 2 (which nobody else has ever mentioned) is probably due to issue 1 or to configuration errors. Issue 3 is normal, and it depends on the default position that the game chooses for the camera in each car. There are plenty of other mods out there that allow you to customize the camera position in every single car model, but it's beyond the scope for me. If it bothers you too much, you can either use another mod to tweak the car parameters, or use convertibles which are great fun to drive :-) Issue 4 is also normal, but in your case it's probably exacerbated by the frame rate problems.

That leaves us with issue 1. Recently it seems that OTT and/or the debug tool have some problems controlling ASW. The only 100% reliable way that I know of to make sure that ASW is off is using the keyboard shortcut Ctrl+NUMPAD1. You can use the Oculus debug HUD to check that ASW is actually off, and not forcing half frame rate. The mod never limits the frame rate artificially, and 40 is too much of a special number, so all clues make me think that your ASW is turning on for some reason.

**Comment by aaronth07** [comment]
>Well, that's a lot of information! First of all, don't start messing with the graphics settings before reading the README. For instance, you're talking about setting Frame Scaling to 0.5 or 0.75 when here I clearly state that it must be 2.5 or 2 for the mod to work correctly.

I only set it to 0.5/0.75 for testing purposes, to make sure it's not a GPU bottleneck. As I said at first, all of these issues happen with my customized settings (which is just enabling AO, dropping frame scaling from 2.5 to 2.0, and raising texture quality), your default medium settings, and my testing settings (which vary).

>So, issue 2 (which nobody else has ever mentioned) is probably due to issue 1 or to configuration errors. 

Hmm, ok. I've never considered that they wouldn't be different issues, they really don't seem related.

>Issue 4 is also normal, but in your case it's probably exacerbated by the frame rate problems.

Oh wow really? I'm actually really surprised at this response considering it is quite distracting. I should have mentioned it also happens even with one eye closed (in-case you might think that it's an artifact of the alternate eye rendering). If it's still intended, well that is a bit unfortunate. But I hope you're correct and it's only more noticeable due to issue 1.

>That leaves us with issue 1. Recently it seems that OTT and/or the debug tool have some problems controlling ASW. The only 100% reliable way that I know of to make sure that ASW is off is using the keyboard shortcut Ctrl+NUMPAD1. You can use the Oculus debug HUD to check that ASW is actually off, and not forcing half frame rate. The mod never limits the frame rate artificially, and 40 is too much of a special number, so all clues make me think that your ASW is turning on for some reason.

Yeah I have had some issues with that before. I can definitely turn off all spacewarp effects (it is not smooth), but I guess not the 40 FPS lock. To use the keyboard shortcut, all I need to have open is your mod? Or OTT? Or the debug tool?

**Comment by LukeRoss00** [comment]
The Ctrl+NUMPAD1 has always been an Oculus feature. As long as the Oculus service is active it should be recognized at all times. But as I said, double check in the HUD: it must say that ASW is not available for the mod to work correctly.

As for issue 4, you can choose to have the rest of the world ghosting when you turn (instead of your arm with the gun) by using the undocumented `~` shortcut when the mod hotkeys are active. You can't fix both though, because the VR runtime can only compensate a single angular velocity, not two independent rotations. I'll tell you, nobody else was ever bothered by the gun ghosting that I remember of, so it must be made much worse by the low frame rate, or you're just very sensitive to it.

**Comment by aaronth07** [comment]
>The Ctrl+NUMPAD1 has always been an Oculus feature. As long as the Oculus service is active it should be recognized at all times. But as I said, double check in the HUD: it must say that ASW is not available for the mod to work correctly.

I'll try that with the HUD enabled, thanks.

>As for issue 4, you can choose to have the rest of the world ghosting when you turn (instead of your arm with the gun) by using the undocumented ~ shortcut when the mod hotkeys are active. You can't fix both though, because the VR runtime can only compensate a single angular velocity, not two independent rotations. I'll tell you, nobody else was ever bothered by the gun ghosting that I remember of, so it must be made much worse by the low frame rate, or you're just very sensitive to it.

I'm not sure. When I say ghosting, I mean serious ghosting on the sides of the weapon. It basically triples the width of the weapon, although of course it's translucent, but that should give you an idea of how extreme the ghosting is. I'll try that hotkey, but I think I would prefer the weapon ghosting over the world ghosting.

**Comment by aaronth07** [comment]
So that seems to work. At least, the HUD said ASW was unavailable. It didn't help much with the ghosting but I would rather have it on the weapon than the world. Actually, my first time trying this mod was with the ghosting on the world instead of the weapon (it's why I thought it was a bug, I guess I accidentally hit the hotkey?), and I thought it was pretty bad TBH, unplayable even. It's definitely much better now, it's going to be great to replay GTA V in VR!

Although, one more question, when disabling ASW using CTRL + Num1, do I have to do that every time I launch the game? And if not, then how do I re-enable it for other games?

**Comment by LukeRoss00** [comment]
Ctrl+Numpad 4 to restore ASW: https://forums.oculusvr.com/t5/Support/What-are-the-keys-for-ASW/m-p/588409/highlight/true#M65887

It's best if you do the Ctrl+Numpad 1 dance every time you start the game. When I was playing I used to have an OTT profile to toggle ASW automatically, but recently it doesn't seem to work reliably anymore (perhaps just on my system, but still).
